# Import Libraries



In [1]:
import pandas as pd
import numpy as np

# Loading data



Column Descriptions:
* Row ID: Unique identifier for each row in the dataset.
*Order ID:Unique identifier for each order placed.
*Order Date: The date when the order was placed.
*Ship Date: The date when the order was shipped.
*Ship Mode: The mode of shipping selected for the order (e.g., *Standard Class, Second Class, First Class, Same Day).
*Customer ID:Unique identifier for each customer.
*Customer Name:The full name of the customer who placed the order.
*Segment:The market segment to which the customer belongs (e.g., Consumer, Corporate, Home Office).
*Postal Code:The postal code for the customer’s location. Note that this column has a significant number of missing values.
* City: The city where the customer is located.
*State:The state where the customer is located.
* Country:The country where the customer is located.
* Region:The region within the country where the customer is located (e.g., East, West, Central, South).
* Market: The broader market category, possibly indicating global regions
*Product ID: Unique identifier for each product.
* Category:The high-level category to which the product belongs.
* Sub-Category:The sub-category of the product .
*Product Name:The name of the product.
* Sales: The sales revenue generated from the product.
* Quantity:The number of units of the product sold.
* Discount: The discount applied to the product, expressed as a decimal (e.g., 0.2 for 20% off).
* Profit: The profit made from the sale of the product.
*Shipping Cost:The cost of shipping the order.
* Order Priority: The priority level of the order (e.g., Low, Medium, High, Critical).





In [93]:
data=pd.read_excel('global_superstore.xlsx')

In [94]:
data

Row ID                  Order ID Order Date  Ship Date       Ship Mode  \
0       40098  CA-2014-AB10015140-41954 2014-11-11 2014-11-13     First Class   
1       26341    IN-2014-JR162107-41675 2014-02-05 2014-02-07    Second Class   
2       25330    IN-2014-CR127307-41929 2014-10-17 2014-10-18     First Class   
3       13524   ES-2014-KM1637548-41667 2014-01-28 2014-01-30     First Class   
4       47221   SG-2014-RH9495111-41948 2014-11-05 2014-11-06        Same Day   
...       ...                       ...        ...        ...             ...   
51285   29002   IN-2015-KE1642066-42174 2015-06-19 2015-06-19        Same Day   
51286   34337  US-2014-ZD21925140-41765 2014-05-06 2014-05-10  Standard Class   
51287   31315  CA-2012-ZD21925140-41147 2012-08-26 2012-08-31    Second Class   
51288    9596   MX-2013-RB1979518-41322 2013-02-17 2013-02-21  Standard Class   
51289    6147   MX-2013-MC1810093-41416 2013-05-22 2013-05-26    Second Class   

        Customer ID       Customer Name      Segment  Postal Code  \
0      AB-100151402       Aaron Bergman     Consumer      73120.0   
1         JR-162107       Justin Ritter    Corporate          NaN   
2         CR-127307        Craig Reiter     Consumer          NaN   
3        KM-1637548    Katherine Murray  Home Office          NaN   
4        RH-9495111         Rick Hansen     Consumer          NaN   
...             ...                 ...          ...          ...   
51285    KE-1642066     Katrina Edelman    Corporate          NaN   
51286  ZD-219251408  Zuschuss Donatelli     Consumer      37421.0   
51287  ZD-219251404  Zuschuss Donatelli     Consumer      94109.0   
51288    RB-1979518          Ross Baird  Home Office          NaN   
51289    MC-1810093       Mick Crebagga     Consumer          NaN   

                City  ...   Product ID         Category Sub-Category  \
0      Oklahoma City  ...  TEC-PH-5816       Technology       Phones   
1         Wollongong  ...  FUR-CH-5379        Furniture       Chairs   
2           Brisbane  ...  TEC-PH-5356       Technology       Phones   
3             Berlin  ...  TEC-PH-5267       Technology       Phones   
4              Dakar  ...  TEC-CO-6011       Technology      Copiers   
...              ...  ...          ...              ...          ...   
51285           Kure  ...  OFF-FA-3072  Office Supplies    Fasteners   
51286    Chattanooga  ...  FUR-FU-4070        Furniture  Furnishings   
51287  San Francisco  ...  OFF-AR-5321  Office Supplies          Art   
51288       Valinhos  ...  OFF-BI-2919  Office Supplies      Binders   
51289       Tipitapa  ...  OFF-PA-3990  Office Supplies        Paper   

                                        Product Name     Sales Quantity  \
0                                   Samsung Convoy 3   221.980        2   
1          Novimex Executive Leather Armchair, Black  3709.395        9   
2                  Nokia Smart Phone, with Caller ID  5175.171        9   
3                     Motorola Smart Phone, Cordless  2892.510        5   
4                     Sharp Wireless Fax, High-Speed  2832.960        8   
...                                              ...       ...      ...   
51285                  Advantus Thumb Tacks, 12 Pack    65.100        5   
51286  Eldon Image Series Desk Accessories, Burgundy    16.720        5   
51287                                     Newell 341     8.560        2   
51288                        Acco Index Tab, Economy    13.440        2   
51289        Eaton Computer Printout Paper, 8.5 x 11    61.380        3   

      Discount    Profit  Shipping Cost  Order Priority  
0          0.0   62.1544         40.770            High  
1          0.1 -288.7650        923.630        Critical  
2          0.1  919.9710        915.490          Medium  
3          0.1  -96.5400        910.160          Medium  
4          0.0  311.5200        903.040        Critical  
...        ...       ...            ...             ...  
51285      0.0    4.5000         

# Exploratory Data Info

In [95]:
data.head(10)

Row ID                  Order ID Order Date  Ship Date       Ship Mode  \
0   40098  CA-2014-AB10015140-41954 2014-11-11 2014-11-13     First Class   
1   26341    IN-2014-JR162107-41675 2014-02-05 2014-02-07    Second Class   
2   25330    IN-2014-CR127307-41929 2014-10-17 2014-10-18     First Class   
3   13524   ES-2014-KM1637548-41667 2014-01-28 2014-01-30     First Class   
4   47221   SG-2014-RH9495111-41948 2014-11-05 2014-11-06        Same Day   
5   22732    IN-2014-JM156557-41818 2014-06-28 2014-07-01    Second Class   
6   30570   IN-2012-TS2134092-41219 2012-11-06 2012-11-08     First Class   
7   31192   IN-2013-MB1808592-41378 2013-04-14 2013-04-18  Standard Class   
8   40099  CA-2014-AB10015140-41954 2014-11-11 2014-11-13     First Class   
9   36258  CA-2012-AB10015140-40974 2012-03-06 2012-03-07     First Class   

    Customer ID     Customer Name      Segment  Postal Code           City  \
0  AB-100151402     Aaron Bergman     Consumer      73120.0  Oklahoma City   
1     JR-162107     Justin Ritter    Corporate          NaN     Wollongong   
2     CR-127307      Craig Reiter     Consumer          NaN       Brisbane   
3    KM-1637548  Katherine Murray  Home Office          NaN         Berlin   
4    RH-9495111       Rick Hansen     Consumer          NaN          Dakar   
5     JM-156557       Jim Mitchum    Corporate          NaN         Sydney   
6    TS-2134092     Toby Swindell     Consumer          NaN        Porirua   
7    MB-1808592        Mick Brown     Consumer          NaN       Hamilton   
8  AB-100151402     Aaron Bergman     Consumer      73120.0  Oklahoma City   
9  AB-100151404     Aaron Bergman     Consumer      98103.0        Seattle   

   ...   Product ID    Category Sub-Category  \
0  ...  TEC-PH-5816  Technology       Phones   
1  ...  FUR-CH-5379   Furniture       Chairs   
2  ...  TEC-PH-5356  Technology       Phones   
3  ...  TEC-PH-5267  Technology       Phones   
4  ...  TEC-CO-6011  Technology      Copiers   
5  ...  TEC-PH-5842  Technology       Phones   
6  ...  FUR-CH-5378   Furniture       Chairs   
7  ...  FUR-TA-3764   Furniture       Tables   
8  ...  FUR-BO-5957   Furniture    Bookcases   
9  ...  FUR-CH-4421   Furniture       Chairs   

                                        Product Name     Sales Quantity  \
0                                   Samsung Convoy 3   221.980        2   
1          Novimex Executive Leather Armchair, Black  3709.395        9   
2                  Nokia Smart Phone, with Caller ID  5175.171        9   
3                     Motorola Smart Phone, Cordless  2892.510        5   
4                     Sharp Wireless Fax, High-Speed  2832.960        8   
5                Samsung Smart Phone, with Caller ID  2862.675        5   
6     Novimex Executive Leather Armchair, Adjustable  1822.080        4   
7       Chromcraft Conference Table, Fully Assembled  5244.840        6   
8  Sauder Facets Collection Library, Sky Alder Fi...   341.960        2   
9         Global Push Button Manager's Chair, Indigo    48.712        1   

  Discount    Profit  Shipping Cost  Order Priority  
0      0.0   62.1544          40.77            High  
1      0.1 -288.7650         923.63        Critical  
2      0.1  919.9710         915.49          Medium  
3      0.1  -96.5400         910.16          Medium  
4      0.0  311.5200         903.04        Critical  
5      0.1  763.2750         897.35        Critical  
6      0.0  564.8400         894.77        Critical  
7      0.0  996.4800         878.38            High  
8      0.0   54.7136          25.27            High  
9      0.2    5.4801          11.13            High  

[10 rows x 24 columns]

In [96]:
data.tail(10)

Row ID                  Order ID Order Date  Ship Date       Ship Mode  \
51280   35112  CA-2014-ZD21925140-41829 2014-07-09 2014-07-09        Same Day   
51281    6039   MX-2015-HG1502518-42164 2015-06-09 2015-06-11     First Class   
51282   24175    IN-2015-DB132707-42221 2015-08-05 2015-08-10  Standard Class   
51283   24105   IN-2015-KH1633058-42154 2015-05-30 2015-05-30        Same Day   
51284    9922   MX-2013-KM1637593-41636 2013-12-28 2013-12-31     First Class   
51285   29002   IN-2015-KE1642066-42174 2015-06-19 2015-06-19        Same Day   
51286   34337  US-2014-ZD21925140-41765 2014-05-06 2014-05-10  Standard Class   
51287   31315  CA-2012-ZD21925140-41147 2012-08-26 2012-08-31    Second Class   
51288    9596   MX-2013-RB1979518-41322 2013-02-17 2013-02-21  Standard Class   
51289    6147   MX-2013-MC1810093-41416 2013-05-22 2013-05-26    Second Class   

        Customer ID       Customer Name      Segment  Postal Code  \
51280  ZD-219251408  Zuschuss Donatelli     Consumer      32216.0   
51281    HG-1502518       Hunter Glantz     Consumer          NaN   
51282     DB-132707   Deborah Brumfield  Home Office          NaN   
51283    KH-1633058     Katharine Harms    Corporate          NaN   
51284    KM-1637593    Katherine Murray  Home Office          NaN   
51285    KE-1642066     Katrina Edelman    Corporate          NaN   
51286  ZD-219251408  Zuschuss Donatelli     Consumer      37421.0   
51287  ZD-219251404  Zuschuss Donatelli     Consumer      94109.0   
51288    RB-1979518          Ross Baird  Home Office          NaN   
51289    MC-1810093       Mick Crebagga     Consumer          NaN   

                    City  ...   Product ID         Category Sub-Category  \
51280       Jacksonville  ...  OFF-PA-6474  Office Supplies        Paper   
51281  Bragança Paulista  ...  OFF-PA-4475  Office Supplies        Paper   
51282         Townsville  ...  OFF-BI-3253  Office Supplies      Binders   
51283            Lucknow  ...  OFF-PA-4007  Office Supplies        Paper   
51284            Managua  ...  OFF-PA-5876  Office Supplies        Paper   
51285               Kure  ...  OFF-FA-3072  Office Supplies    Fasteners   
51286        Chattanooga  ...  FUR-FU-4070        Furniture  Furnishings   
51287      San Francisco  ...  OFF-AR-5321  Office Supplies          Art   
51288           Valinhos  ...  OFF-BI-2919  Office Supplies      Binders   
51289           Tipitapa  ...  OFF-PA-3990  Office Supplies        Paper   

                                        Product Name   Sales Quantity  \
51280                                     Xerox 1921  15.984        2   
51281            Green Bar Message Books, Multicolor  84.000        5   
51282                          Avery Binder, Economy  58.050        5   
51283                 Eaton Parchment Paper, Premium  26.940        2   
51284                SanDisk Message Books, 8.5 x 11  18.640        1   
51285                  Advantus Thumb Tacks, 12 Pack  65.100        5   
51286  Eldon Image Series Desk Accessories, Burgundy  16.720        5   
51287                                     Newell 341   8.560        2   
51288                        Acco Index Tab, Economy  13.440        2   
51289        Eaton Computer Printout Paper, 8.5 x 11  61.380        3   

      Discount   Profit  Shipping Cost  Order Priority  
51280      0.2   4.9950          2.010          Medium  
51281      0.0   9.2000          1.019            High  
51282      0.1  19.9500          1.010          Medium  
51283      0.0   1.8600          1.010            High  
51284      0.0   8.0000          1.010          Medium  
51285      0.0   4.5000          1.010          Medium  
51286      0.2   3.3440          1.930            High  
51287      0.0   2.4824          1.580            High  
51288      0.0   2.4000          1.003          Medium  
51289      0.0   1.8000          1.002            High  

[10 rows x 24 columns]

In [97]:
data.shape

(51290, 24)

In [98]:
data.size

1230960

In [99]:
51290*24

1230960

In [100]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51290 entries, 0 to 51289
Data columns (total 24 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   Row ID          51290 non-null  int64         
 1   Order ID        51290 non-null  object        
 2   Order Date      51290 non-null  datetime64[ns]
 3   Ship Date       51290 non-null  datetime64[ns]
 4   Ship Mode       51290 non-null  object        
 5   Customer ID     51290 non-null  object        
 6   Customer Name   51290 non-null  object        
 7   Segment         51290 non-null  object        
 8   Postal Code     9994 non-null   float64       
 9   City            51290 non-null  object        
 10  State           51290 non-null  object        
 11  Country         51290 non-null  object        
 12  Region          51290 non-null  object        
 13  Market          51290 non-null  object        
 14  Product ID      51290 non-null  object        
 15  Ca



*   The dataset has 15 categorical columns (e.g., 'Order ID', 'Customer Name', 'Region') and 7 numerical columns (e.g., 'Sales', 'Quantity', 'Discount').

*   Two datetime columns ('Order Date', 'Ship Date') are crucial for time-series analysis.
* The 'Postal Code' column has a significant amount of missing
data, with only 9,994 out of 51,290 entries being non-null. This means approximately 80% of the data is missing in this column.



In [101]:
data.isnull().sum()

Row ID                0
Order ID              0
Order Date            0
Ship Date             0
Ship Mode             0
Customer ID           0
Customer Name         0
Segment               0
Postal Code       41296
City                  0
State                 0
Country               0
Region                0
Market                0
Product ID            0
Category              0
Sub-Category          0
Product Name          0
Sales                 0
Quantity              0
Discount              0
Profit                0
Shipping Cost         0
Order Priority        0
dtype: int64

In [102]:
data.describe()

Row ID                     Order Date  \
count  51290.00000                          51290   
mean   25645.50000  2014-05-11 21:26:49.155780608   
min        1.00000            2012-01-01 00:00:00   
25%    12823.25000            2013-06-19 00:00:00   
50%    25645.50000            2014-07-08 00:00:00   
75%    38467.75000            2015-05-22 00:00:00   
max    51290.00000            2015-12-31 00:00:00   
std    14806.29199                            NaN   

                           Ship Date   Postal Code         Sales  \
count                          51290   9994.000000  51290.000000   
mean   2014-05-15 20:42:42.745174528  55190.379428    246.490581   
min              2012-01-03 00:00:00   1040.000000      0.444000   
25%              2013-06-23 00:00:00  23223.000000     30.758625   
50%              2014-07-12 00:00:00  56430.500000     85.053000   
75%              2015-05-26 00:00:00  90008.000000    251.053200   
max              2016-01-07 00:00:00  99301.000000  22638.480000   
std                              NaN  32063.693350    487.565361   

           Quantity      Discount        Profit  Shipping Cost  
count  51290.000000  51290.000000  51290.000000   51290.000000  
mean       3.476545      0.142908     28.610982      26.478567  
min        1.000000      0.000000  -6599.978000       1.002000  
25%        2.000000      0.000000      0.000000       2.610000  
50%        3.000000      0.000000      9.240000       7.790000  
75%        5.000000      0.200000     36.810000      24.450000  
max       14.000000      0.850000   8399.976000     933.570000  
std        2.278766      0.212280    174.340972      57.251373

In [103]:
data.describe(include='all').T

count unique                       top   freq  \
Row ID          51290.0    NaN                       NaN    NaN   
Order ID          51290  25728  CA-2015-SV20365140-42268     14   
Order Date        51290    NaN                       NaN    NaN   
Ship Date         51290    NaN                       NaN    NaN   
Ship Mode         51290      4            Standard Class  30775   
Customer ID       51290  17415              SV-203651406     26   
Customer Name     51290    796           Muhammed Yedwab    108   
Segment           51290      3                  Consumer  26518   
Postal Code      9994.0    NaN                       NaN    NaN   
City              51290   3650             New York City    915   
State             51290   1102                California   2001   
Country           51290    165             United States   9994   
Region            51290     23            Western Europe   5883   
Market            51290      5              Asia Pacific  14302   
Product ID        51290   3788               OFF-FA-6129    227   
Category          51290      3           Office Supplies  31289   
Sub-Category      51290     17                   Binders   6146   
Product Name      51290   3788                   Staples    227   
Sales           51290.0    NaN                       NaN    NaN   
Quantity        51290.0    NaN                       NaN    NaN   
Discount        51290.0    NaN                       NaN    NaN   
Profit          51290.0    NaN                       NaN    NaN   
Shipping Cost   51290.0    NaN                       NaN    NaN   
Order Priority    51290      4                    Medium  29433   

                                         mean                  min  \
Row ID                                25645.5                  1.0   
Order ID                                  NaN                  NaN   
Order Date      2014-05-11 21:26:49.155780608  2012-01-01 00:00:00   
Ship Date       2014-05-15 20:42:42.745174528  2012-01-03 00:00:00   
Ship Mode                                 NaN                  NaN   
Customer ID                               NaN                  NaN   
Customer Name                             NaN                  NaN   
Segment                                   NaN                  NaN   
Postal Code                      55190.379428               1040.0   
City                                      NaN                  NaN   
State                                     NaN                  NaN   
Country                                   NaN                  NaN   
Region                                    NaN                  NaN   
Market                                    NaN                  NaN   
Product ID                                NaN                  NaN   
Category                                  NaN                  NaN   
Sub-Category                              NaN                  NaN   
Product Name                              NaN                  NaN   
Sales                              246.490581                0.444   
Quantity                             3.476545                  1.0   
Discount                             0.142908                  0.0   
Profit                              28.610982            -6599.978   
Shipping Cost                       26.478567                1.002   
Order Priority                            NaN                  NaN   

                                25%                  50%                  75%  \
Row ID                     12823.25              25645.5             38467.75   
Order ID                        NaN                  NaN                  NaN   
Order Date      2013-06-19 00:00:00  2014-07-08 00:00:00  2015-05-22 00:00:00   
Ship Date       2013-06-23 00:00:00  2014-07-12 00:00:00  2015-05-26 00:00:00   
Ship Mode                       NaN                  NaN                  NaN   
Customer ID                     NaN                  NaN                  NaN   
Customer Name           

In [104]:
data.describe(include='object')

Order ID       Ship Mode   Customer ID  \
count                      51290           51290         51290   
unique                     25728               4         17415   
top     CA-2015-SV20365140-42268  Standard Class  SV-203651406   
freq                          14           30775            26   

          Customer Name   Segment           City       State        Country  \
count             51290     51290          51290       51290          51290   
unique              796         3           3650        1102            165   
top     Muhammed Yedwab  Consumer  New York City  California  United States   
freq                108     26518            915        2001           9994   

                Region        Market   Product ID         Category  \
count            51290         51290        51290            51290   
unique              23             5         3788                3   
top     Western Europe  Asia Pacific  OFF-FA-6129  Office Supplies   
freq              5883         14302          227            31289   

       Sub-Category Product Name Order Priority  
count         51290        51290          51290  
unique           17         3788              4  
top         Binders      Staples         Medium  
freq           6146          227          29433

In [105]:
data.nunique()

Row ID            51290
Order ID          25728
Order Date         1430
Ship Date          1464
Ship Mode             4
Customer ID       17415
Customer Name       796
Segment               3
Postal Code         631
City               3650
State              1102
Country             165
Region               23
Market                5
Product ID         3788
Category              3
Sub-Category         17
Product Name       3788
Sales             27200
Quantity             14
Discount             29
Profit            28234
Shipping Cost     16753
Order Priority        4
dtype: int64

# Checking Duplicates

In [106]:
data.duplicated()

0        False
1        False
2        False
3        False
4        False
         ...  
51285    False
51286    False
51287    False
51288    False
51289    False
Length: 51290, dtype: bool

In [107]:
data

Row ID                  Order ID Order Date  Ship Date       Ship Mode  \
0       40098  CA-2014-AB10015140-41954 2014-11-11 2014-11-13     First Class   
1       26341    IN-2014-JR162107-41675 2014-02-05 2014-02-07    Second Class   
2       25330    IN-2014-CR127307-41929 2014-10-17 2014-10-18     First Class   
3       13524   ES-2014-KM1637548-41667 2014-01-28 2014-01-30     First Class   
4       47221   SG-2014-RH9495111-41948 2014-11-05 2014-11-06        Same Day   
...       ...                       ...        ...        ...             ...   
51285   29002   IN-2015-KE1642066-42174 2015-06-19 2015-06-19        Same Day   
51286   34337  US-2014-ZD21925140-41765 2014-05-06 2014-05-10  Standard Class   
51287   31315  CA-2012-ZD21925140-41147 2012-08-26 2012-08-31    Second Class   
51288    9596   MX-2013-RB1979518-41322 2013-02-17 2013-02-21  Standard Class   
51289    6147   MX-2013-MC1810093-41416 2013-05-22 2013-05-26    Second Class   

        Customer ID       Customer Name      Segment  Postal Code  \
0      AB-100151402       Aaron Bergman     Consumer      73120.0   
1         JR-162107       Justin Ritter    Corporate          NaN   
2         CR-127307        Craig Reiter     Consumer          NaN   
3        KM-1637548    Katherine Murray  Home Office          NaN   
4        RH-9495111         Rick Hansen     Consumer          NaN   
...             ...                 ...          ...          ...   
51285    KE-1642066     Katrina Edelman    Corporate          NaN   
51286  ZD-219251408  Zuschuss Donatelli     Consumer      37421.0   
51287  ZD-219251404  Zuschuss Donatelli     Consumer      94109.0   
51288    RB-1979518          Ross Baird  Home Office          NaN   
51289    MC-1810093       Mick Crebagga     Consumer          NaN   

                City  ...   Product ID         Category Sub-Category  \
0      Oklahoma City  ...  TEC-PH-5816       Technology       Phones   
1         Wollongong  ...  FUR-CH-5379        Furniture       Chairs   
2           Brisbane  ...  TEC-PH-5356       Technology       Phones   
3             Berlin  ...  TEC-PH-5267       Technology       Phones   
4              Dakar  ...  TEC-CO-6011       Technology      Copiers   
...              ...  ...          ...              ...          ...   
51285           Kure  ...  OFF-FA-3072  Office Supplies    Fasteners   
51286    Chattanooga  ...  FUR-FU-4070        Furniture  Furnishings   
51287  San Francisco  ...  OFF-AR-5321  Office Supplies          Art   
51288       Valinhos  ...  OFF-BI-2919  Office Supplies      Binders   
51289       Tipitapa  ...  OFF-PA-3990  Office Supplies        Paper   

                                        Product Name     Sales Quantity  \
0                                   Samsung Convoy 3   221.980        2   
1          Novimex Executive Leather Armchair, Black  3709.395        9   
2                  Nokia Smart Phone, with Caller ID  5175.171        9   
3                     Motorola Smart Phone, Cordless  2892.510        5   
4                     Sharp Wireless Fax, High-Speed  2832.960        8   
...                                              ...       ...      ...   
51285                  Advantus Thumb Tacks, 12 Pack    65.100        5   
51286  Eldon Image Series Desk Accessories, Burgundy    16.720        5   
51287                                     Newell 341     8.560        2   
51288                        Acco Index Tab, Economy    13.440        2   
51289        Eaton Computer Printout Paper, 8.5 x 11    61.380        3   

      Discount    Profit  Shipping Cost  Order Priority  
0          0.0   62.1544         40.770            High  
1          0.1 -288.7650        923.630        Critical  
2          0.1  919.9710        915.490          Medium  
3          0.1  -96.5400        910.160          Medium  
4          0.0  311.5200        903.040        Critical  
...        ...       ...            ...             ...  
51285      0.0    4.5000         

In [108]:
data.columns

Index(['Row ID', 'Order ID', 'Order Date', 'Ship Date', 'Ship Mode',
       'Customer ID', 'Customer Name', 'Segment', 'Postal Code', 'City',
       'State', 'Country', 'Region', 'Market', 'Product ID', 'Category',
       'Sub-Category', 'Product Name', 'Sales', 'Quantity', 'Discount',
       'Profit', 'Shipping Cost', 'Order Priority'],
      dtype='object')

In [109]:
data.duplicated().sum()

0

How many times a specific customer ('Customer ID') has purchased the same product ('Product ID') more than once.

In [110]:
data.duplicated(subset=['Customer ID','Product ID']).sum()

118

In [111]:
df1=data[data.duplicated(subset=['Customer ID','Product ID'])]

In [112]:
df1

Row ID                  Order ID Order Date  Ship Date       Ship Mode  \
143     31649  CA-2014-AB10060140-41884 2014-09-02 2014-09-04     First Class   
2628    40081  US-2015-AS10630140-42081 2015-03-18 2015-03-24  Standard Class   
4257    33490  CA-2014-AP10915140-41984 2014-12-11 2014-12-18  Standard Class   
4892    10287   US-2014-DN1369018-41688 2014-02-18 2014-02-22  Standard Class   
5091     2377   MX-2013-AG1090082-41398 2013-05-04 2013-05-09  Standard Class   
...       ...                       ...        ...        ...             ...   
50428   10001   US-2014-DP1300018-41909 2014-09-27 2014-10-01  Standard Class   
50611      11   MX-2014-DP1300018-41909 2014-09-27 2014-10-01  Standard Class   
50963   38356  CA-2014-WB21850140-41985 2014-12-12 2014-12-12        Same Day   
51017   10091   US-2012-KT1646518-41017 2012-04-18 2012-04-24  Standard Class   
51129   37507  CA-2015-YS21880140-42235 2015-08-19 2015-08-26  Standard Class   

        Customer ID    Customer Name      Segment  Postal Code           City  \
143    AB-100601406  Adam Bellavance  Home Office      10009.0  New York City   
2628   AS-106301408       Ann Steele  Home Office      37918.0      Knoxville   
4257   AP-109151404   Arthur Prichep     Consumer      98105.0        Seattle   
4892     DN-1369018     Duane Noonan     Consumer          NaN            Açu   
5091     AG-1090082    Arthur Gainer     Consumer          NaN    Mexico City   
...             ...              ...          ...          ...            ...   
50428    DP-1300018    Darren Powers     Consumer          NaN      Araranguá   
50611    DP-1300018    Darren Powers     Consumer          NaN      São Paulo   
50963  WB-218501404    William Brown     Consumer      92804.0        Anaheim   
51017    KT-1646518    Kean Takahito     Consumer          NaN      Araranguá   
51129  YS-218801404    Yana Sorensen    Corporate      98103.0        Seattle   

       ...   Product ID         Category Sub-Category  \
143    ...  OFF-PA-6430  Office Supplies        Paper   
2628   ...  OFF-LA-3192  Office Supplies       Labels   
4257   ...  OFF-PA-6546  Office Supplies        Paper   
4892   ...  FUR-TA-4645        Furniture       Tables   
5091   ...  OFF-AP-4508  Office Supplies   Appliances   
...    ...          ...              ...          ...   
50428  ...  OFF-EN-4446  Office Supplies    Envelopes   
50611  ...  OFF-BI-6398  Office Supplies      Binders   
50963  ...  OFF-AP-4239  Office Supplies   Appliances   
51017  ...  TEC-AC-4151       Technology  Accessories   
51129  ...  OFF-ST-5076  Office Supplies      Storage   

                                            Product Name     Sales Quantity  \
143                                           Xerox 1881    49.120        4   
2628                                           Avery 474     4.608        2   
4257                                          Xerox 1988   123.920        4   
4892             Hon Conference Table, Adjustable Height   985.248        4   
5091                         Hamilton Beach Stove, White   720.640        2   
...                                                  ...       ...      ...   
50428          GlobeWeis Mailers, with clear poly window    22.448        2   
50611          Wilson Jones Hole Reinforcements, Durable    12.060        3   
50963  Fellowes 8 Outlet Superior Workstation Surge P...    83.420        2   
51017                     Enermax Flash Drive, Bluetooth    22.144        2   
51129                                   Letter Size Cart  1000.020        7   

      Discount    Profit  Shipping Cost  Order Priority  
143        0.0   23.0864          6.220        Critical  
2628       0.2    1.6704          1.220          Medium  
4257       0.0   55.7640         12.440             Low  
4892       0.6 -936.0320         67.853            High  
5091       0.0   72.0400         65.580          Medium  
...        ...       ...            ...             ...  
50428      0.6  -12.3

In [113]:
data

Row ID                  Order ID Order Date  Ship Date       Ship Mode  \
0       40098  CA-2014-AB10015140-41954 2014-11-11 2014-11-13     First Class   
1       26341    IN-2014-JR162107-41675 2014-02-05 2014-02-07    Second Class   
2       25330    IN-2014-CR127307-41929 2014-10-17 2014-10-18     First Class   
3       13524   ES-2014-KM1637548-41667 2014-01-28 2014-01-30     First Class   
4       47221   SG-2014-RH9495111-41948 2014-11-05 2014-11-06        Same Day   
...       ...                       ...        ...        ...             ...   
51285   29002   IN-2015-KE1642066-42174 2015-06-19 2015-06-19        Same Day   
51286   34337  US-2014-ZD21925140-41765 2014-05-06 2014-05-10  Standard Class   
51287   31315  CA-2012-ZD21925140-41147 2012-08-26 2012-08-31    Second Class   
51288    9596   MX-2013-RB1979518-41322 2013-02-17 2013-02-21  Standard Class   
51289    6147   MX-2013-MC1810093-41416 2013-05-22 2013-05-26    Second Class   

        Customer ID       Customer Name      Segment  Postal Code  \
0      AB-100151402       Aaron Bergman     Consumer      73120.0   
1         JR-162107       Justin Ritter    Corporate          NaN   
2         CR-127307        Craig Reiter     Consumer          NaN   
3        KM-1637548    Katherine Murray  Home Office          NaN   
4        RH-9495111         Rick Hansen     Consumer          NaN   
...             ...                 ...          ...          ...   
51285    KE-1642066     Katrina Edelman    Corporate          NaN   
51286  ZD-219251408  Zuschuss Donatelli     Consumer      37421.0   
51287  ZD-219251404  Zuschuss Donatelli     Consumer      94109.0   
51288    RB-1979518          Ross Baird  Home Office          NaN   
51289    MC-1810093       Mick Crebagga     Consumer          NaN   

                City  ...   Product ID         Category Sub-Category  \
0      Oklahoma City  ...  TEC-PH-5816       Technology       Phones   
1         Wollongong  ...  FUR-CH-5379        Furniture       Chairs   
2           Brisbane  ...  TEC-PH-5356       Technology       Phones   
3             Berlin  ...  TEC-PH-5267       Technology       Phones   
4              Dakar  ...  TEC-CO-6011       Technology      Copiers   
...              ...  ...          ...              ...          ...   
51285           Kure  ...  OFF-FA-3072  Office Supplies    Fasteners   
51286    Chattanooga  ...  FUR-FU-4070        Furniture  Furnishings   
51287  San Francisco  ...  OFF-AR-5321  Office Supplies          Art   
51288       Valinhos  ...  OFF-BI-2919  Office Supplies      Binders   
51289       Tipitapa  ...  OFF-PA-3990  Office Supplies        Paper   

                                        Product Name     Sales Quantity  \
0                                   Samsung Convoy 3   221.980        2   
1          Novimex Executive Leather Armchair, Black  3709.395        9   
2                  Nokia Smart Phone, with Caller ID  5175.171        9   
3                     Motorola Smart Phone, Cordless  2892.510        5   
4                     Sharp Wireless Fax, High-Speed  2832.960        8   
...                                              ...       ...      ...   
51285                  Advantus Thumb Tacks, 12 Pack    65.100        5   
51286  Eldon Image Series Desk Accessories, Burgundy    16.720        5   
51287                                     Newell 341     8.560        2   
51288                        Acco Index Tab, Economy    13.440        2   
51289        Eaton Computer Printout Paper, 8.5 x 11    61.380        3   

      Discount    Profit  Shipping Cost  Order Priority  
0          0.0   62.1544         40.770            High  
1          0.1 -288.7650        923.630        Critical  
2          0.1  919.9710        915.490          Medium  
3          0.1  -96.5400        910.160          Medium  
4          0.0  311.5200        903.040        Critical  
...        ...       ...            ...             ...  
51285      0.0    4.5000         

In [114]:
data.drop_duplicates(keep='first',inplace=True)

In [115]:
data_6 = { 'Name': ['Alice', 'Bob', 'Charlie', 'Alice', 'Eve', 'Alice'],
        'Age': [24, 27, 22, 24, 29, 24],
      'City': ['New York', 'Los Angeles', 'Chicago', 'New York', 'Chicago', 'New York'] }
df6=pd.DataFrame(data_6)

In [116]:
df6

Name  Age         City
0    Alice   24     New York
1      Bob   27  Los Angeles
2  Charlie   22      Chicago
3    Alice   24     New York
4      Eve   29      Chicago
5    Alice   24     New York

In [117]:
df6.duplicated()

0    False
1    False
2    False
3     True
4    False
5     True
dtype: bool

In [118]:
df_F=df6.drop_duplicates(keep='first')

In [119]:
df_F

Name  Age         City
0    Alice   24     New York
1      Bob   27  Los Angeles
2  Charlie   22      Chicago
4      Eve   29      Chicago

In [120]:
df_L=df6.drop_duplicates(keep='last')

In [121]:
df_L

Name  Age         City
1      Bob   27  Los Angeles
2  Charlie   22      Chicago
4      Eve   29      Chicago
5    Alice   24     New York

In [122]:
df_=df6.drop_duplicates(keep=False)

In [123]:
df_

Name  Age         City
1      Bob   27  Los Angeles
2  Charlie   22      Chicago
4      Eve   29      Chicago

# Convert Data Types

In [124]:
data.dtypes

Row ID                     int64
Order ID                  object
Order Date        datetime64[ns]
Ship Date         datetime64[ns]
Ship Mode                 object
Customer ID               object
Customer Name             object
Segment                   object
Postal Code              float64
City                      object
State                     object
Country                   object
Region                    object
Market                    object
Product ID                object
Category                  object
Sub-Category              object
Product Name              object
Sales                    float64
Quantity                   int64
Discount                 float64
Profit                   float64
Shipping Cost            float64
Order Priority            object
dtype: object

In [125]:
data['Postal Code']=data['Postal Code'].astype(str)

In [126]:
data.dtypes

Row ID                     int64
Order ID                  object
Order Date        datetime64[ns]
Ship Date         datetime64[ns]
Ship Mode                 object
Customer ID               object
Customer Name             object
Segment                   object
Postal Code               object
City                      object
State                     object
Country                   object
Region                    object
Market                    object
Product ID                object
Category                  object
Sub-Category              object
Product Name              object
Sales                    float64
Quantity                   int64
Discount                 float64
Profit                   float64
Shipping Cost            float64
Order Priority            object
dtype: object

# Feature Engineering

In [127]:
data.columns

Index(['Row ID', 'Order ID', 'Order Date', 'Ship Date', 'Ship Mode',
       'Customer ID', 'Customer Name', 'Segment', 'Postal Code', 'City',
       'State', 'Country', 'Region', 'Market', 'Product ID', 'Category',
       'Sub-Category', 'Product Name', 'Sales', 'Quantity', 'Discount',
       'Profit', 'Shipping Cost', 'Order Priority'],
      dtype='object')

###  Create a new feature that represents the total value of each order

In [128]:
data[['Sales','Quantity']]

Sales  Quantity
0       221.980         2
1      3709.395         9
2      5175.171         9
3      2892.510         5
4      2832.960         8
...         ...       ...
51285    65.100         5
51286    16.720         5
51287     8.560         2
51288    13.440         2
51289    61.380         3

[51290 rows x 2 columns]

In [129]:
data['Total_Order_Value']=data['Sales']*data['Quantity']

In [130]:
data.head(5)

Row ID                  Order ID Order Date  Ship Date     Ship Mode  \
0   40098  CA-2014-AB10015140-41954 2014-11-11 2014-11-13   First Class   
1   26341    IN-2014-JR162107-41675 2014-02-05 2014-02-07  Second Class   
2   25330    IN-2014-CR127307-41929 2014-10-17 2014-10-18   First Class   
3   13524   ES-2014-KM1637548-41667 2014-01-28 2014-01-30   First Class   
4   47221   SG-2014-RH9495111-41948 2014-11-05 2014-11-06      Same Day   

    Customer ID     Customer Name      Segment Postal Code           City  \
0  AB-100151402     Aaron Bergman     Consumer     73120.0  Oklahoma City   
1     JR-162107     Justin Ritter    Corporate         nan     Wollongong   
2     CR-127307      Craig Reiter     Consumer         nan       Brisbane   
3    KM-1637548  Katherine Murray  Home Office         nan         Berlin   
4    RH-9495111       Rick Hansen     Consumer         nan          Dakar   

   ...    Category Sub-Category                               Product Name  \
0  ...  Technology       Phones                           Samsung Convoy 3   
1  ...   Furniture       Chairs  Novimex Executive Leather Armchair, Black   
2  ...  Technology       Phones          Nokia Smart Phone, with Caller ID   
3  ...  Technology       Phones             Motorola Smart Phone, Cordless   
4  ...  Technology      Copiers             Sharp Wireless Fax, High-Speed   

      Sales Quantity Discount    Profit Shipping Cost  Order Priority  \
0   221.980        2      0.0   62.1544         40.77            High   
1  3709.395        9      0.1 -288.7650        923.63        Critical   
2  5175.171        9      0.1  919.9710        915.49          Medium   
3  2892.510        5      0.1  -96.5400        910.16          Medium   
4  2832.960        8      0.0  311.5200        903.04        Critical   

   Total_Order_Value  
0            443.960  
1          33384.555  
2          46576.539  
3          14462.550  
4          22663.680  

[5 rows x 25 columns]

### Create a feature that calculates the profit margin as a percentage of the total sales.


In [131]:
data['profit margin']=(data['Profit']/data['Sales'])*100

In [132]:
data.head(5)

Row ID                  Order ID Order Date  Ship Date     Ship Mode  \
0   40098  CA-2014-AB10015140-41954 2014-11-11 2014-11-13   First Class   
1   26341    IN-2014-JR162107-41675 2014-02-05 2014-02-07  Second Class   
2   25330    IN-2014-CR127307-41929 2014-10-17 2014-10-18   First Class   
3   13524   ES-2014-KM1637548-41667 2014-01-28 2014-01-30   First Class   
4   47221   SG-2014-RH9495111-41948 2014-11-05 2014-11-06      Same Day   

    Customer ID     Customer Name      Segment Postal Code           City  \
0  AB-100151402     Aaron Bergman     Consumer     73120.0  Oklahoma City   
1     JR-162107     Justin Ritter    Corporate         nan     Wollongong   
2     CR-127307      Craig Reiter     Consumer         nan       Brisbane   
3    KM-1637548  Katherine Murray  Home Office         nan         Berlin   
4    RH-9495111       Rick Hansen     Consumer         nan          Dakar   

   ... Sub-Category                               Product Name     Sales  \
0  ...       Phones                           Samsung Convoy 3   221.980   
1  ...       Chairs  Novimex Executive Leather Armchair, Black  3709.395   
2  ...       Phones          Nokia Smart Phone, with Caller ID  5175.171   
3  ...       Phones             Motorola Smart Phone, Cordless  2892.510   
4  ...      Copiers             Sharp Wireless Fax, High-Speed  2832.960   

  Quantity Discount    Profit Shipping Cost Order Priority  Total_Order_Value  \
0        2      0.0   62.1544         40.77           High            443.960   
1        9      0.1 -288.7650        923.63       Critical          33384.555   
2        9      0.1  919.9710        915.49         Medium          46576.539   
3        5      0.1  -96.5400        910.16         Medium          14462.550   
4        8      0.0  311.5200        903.04       Critical          22663.680   

   profit margin  
0      28.000000  
1      -7.784693  
2      17.776630  
3      -3.337586  
4      10.996272  

[5 rows x 26 columns]

### Create a feature that calculates the price of the product after applying the discount.

In [133]:
data['price_after_discount']=data['Sales']* (1-data['Discount'])

In [134]:
data.head(5)

Row ID                  Order ID Order Date  Ship Date     Ship Mode  \
0   40098  CA-2014-AB10015140-41954 2014-11-11 2014-11-13   First Class   
1   26341    IN-2014-JR162107-41675 2014-02-05 2014-02-07  Second Class   
2   25330    IN-2014-CR127307-41929 2014-10-17 2014-10-18   First Class   
3   13524   ES-2014-KM1637548-41667 2014-01-28 2014-01-30   First Class   
4   47221   SG-2014-RH9495111-41948 2014-11-05 2014-11-06      Same Day   

    Customer ID     Customer Name      Segment Postal Code           City  \
0  AB-100151402     Aaron Bergman     Consumer     73120.0  Oklahoma City   
1     JR-162107     Justin Ritter    Corporate         nan     Wollongong   
2     CR-127307      Craig Reiter     Consumer         nan       Brisbane   
3    KM-1637548  Katherine Murray  Home Office         nan         Berlin   
4    RH-9495111       Rick Hansen     Consumer         nan          Dakar   

   ...                               Product Name     Sales Quantity Discount  \
0  ...                           Samsung Convoy 3   221.980        2      0.0   
1  ...  Novimex Executive Leather Armchair, Black  3709.395        9      0.1   
2  ...          Nokia Smart Phone, with Caller ID  5175.171        9      0.1   
3  ...             Motorola Smart Phone, Cordless  2892.510        5      0.1   
4  ...             Sharp Wireless Fax, High-Speed  2832.960        8      0.0   

     Profit Shipping Cost Order Priority Total_Order_Value  profit margin  \
0   62.1544         40.77           High           443.960      28.000000   
1 -288.7650        923.63       Critical         33384.555      -7.784693   
2  919.9710        915.49         Medium         46576.539      17.776630   
3  -96.5400        910.16         Medium         14462.550      -3.337586   
4  311.5200        903.04       Critical         22663.680      10.996272   

   price_after_discount  
0              221.9800  
1             3338.4555  
2             4657.6539  
3             2603.2590  
4             2832.9600  

[5 rows x 27 columns]

### Create a feature that represents the shipping cost per unit sold.

In [135]:
data['Shipping_cost_per_unit']=data['Shipping Cost']/data['Quantity']

# Extracting Date and Time Components
* Extracting date and time components from datetime objects




In [136]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51290 entries, 0 to 51289
Data columns (total 28 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   Row ID                  51290 non-null  int64         
 1   Order ID                51290 non-null  object        
 2   Order Date              51290 non-null  datetime64[ns]
 3   Ship Date               51290 non-null  datetime64[ns]
 4   Ship Mode               51290 non-null  object        
 5   Customer ID             51290 non-null  object        
 6   Customer Name           51290 non-null  object        
 7   Segment                 51290 non-null  object        
 8   Postal Code             51290 non-null  object        
 9   City                    51290 non-null  object        
 10  State                   51290 non-null  object        
 11  Country                 51290 non-null  object        
 12  Region                  51290 non-null  object

### Extract the year from the 'Order Date' to create a new feature indicating the year in which the order was placed.

In [137]:
data['order_year']=data['Order Date'].dt.year

In [138]:
data.head()

Row ID                  Order ID Order Date  Ship Date     Ship Mode  \
0   40098  CA-2014-AB10015140-41954 2014-11-11 2014-11-13   First Class   
1   26341    IN-2014-JR162107-41675 2014-02-05 2014-02-07  Second Class   
2   25330    IN-2014-CR127307-41929 2014-10-17 2014-10-18   First Class   
3   13524   ES-2014-KM1637548-41667 2014-01-28 2014-01-30   First Class   
4   47221   SG-2014-RH9495111-41948 2014-11-05 2014-11-06      Same Day   

    Customer ID     Customer Name      Segment Postal Code           City  \
0  AB-100151402     Aaron Bergman     Consumer     73120.0  Oklahoma City   
1     JR-162107     Justin Ritter    Corporate         nan     Wollongong   
2     CR-127307      Craig Reiter     Consumer         nan       Brisbane   
3    KM-1637548  Katherine Murray  Home Office         nan         Berlin   
4    RH-9495111       Rick Hansen     Consumer         nan          Dakar   

   ... Quantity Discount    Profit Shipping Cost Order Priority  \
0  ...        2      0.0   62.1544         40.77           High   
1  ...        9      0.1 -288.7650        923.63       Critical   
2  ...        9      0.1  919.9710        915.49         Medium   
3  ...        5      0.1  -96.5400        910.16         Medium   
4  ...        8      0.0  311.5200        903.04       Critical   

  Total_Order_Value profit margin price_after_discount  \
0           443.960     28.000000             221.9800   
1         33384.555     -7.784693            3338.4555   
2         46576.539     17.776630            4657.6539   
3         14462.550     -3.337586            2603.2590   
4         22663.680     10.996272            2832.9600   

   Shipping_cost_per_unit  order_year  
0               20.385000        2014  
1              102.625556        2014  
2              101.721111        2014  
3              182.032000        2014  
4              112.880000        2014  

[5 rows x 29 columns]

### Extract the month from the 'Order Date' to create a new feature indicating the month in which the order was placed.

In [139]:
data['order_month']=data['Order Date'].dt.month
data.head()

Row ID                  Order ID Order Date  Ship Date     Ship Mode  \
0   40098  CA-2014-AB10015140-41954 2014-11-11 2014-11-13   First Class   
1   26341    IN-2014-JR162107-41675 2014-02-05 2014-02-07  Second Class   
2   25330    IN-2014-CR127307-41929 2014-10-17 2014-10-18   First Class   
3   13524   ES-2014-KM1637548-41667 2014-01-28 2014-01-30   First Class   
4   47221   SG-2014-RH9495111-41948 2014-11-05 2014-11-06      Same Day   

    Customer ID     Customer Name      Segment Postal Code           City  \
0  AB-100151402     Aaron Bergman     Consumer     73120.0  Oklahoma City   
1     JR-162107     Justin Ritter    Corporate         nan     Wollongong   
2     CR-127307      Craig Reiter     Consumer         nan       Brisbane   
3    KM-1637548  Katherine Murray  Home Office         nan         Berlin   
4    RH-9495111       Rick Hansen     Consumer         nan          Dakar   

   ... Discount    Profit Shipping Cost Order Priority Total_Order_Value  \
0  ...      0.0   62.1544         40.77           High           443.960   
1  ...      0.1 -288.7650        923.63       Critical         33384.555   
2  ...      0.1  919.9710        915.49         Medium         46576.539   
3  ...      0.1  -96.5400        910.16         Medium         14462.550   
4  ...      0.0  311.5200        903.04       Critical         22663.680   

  profit margin price_after_discount Shipping_cost_per_unit  order_year  \
0     28.000000             221.9800              20.385000        2014   
1     -7.784693            3338.4555             102.625556        2014   
2     17.776630            4657.6539             101.721111        2014   
3     -3.337586            2603.2590             182.032000        2014   
4     10.996272            2832.9600             112.880000        2014   

   order_month  
0           11  
1            2  
2           10  
3            1  
4           11  

[5 rows x 30 columns]

In [140]:
data['order_month_name']=data['Order Date'].dt.month_name()
data.head()

Row ID                  Order ID Order Date  Ship Date     Ship Mode  \
0   40098  CA-2014-AB10015140-41954 2014-11-11 2014-11-13   First Class   
1   26341    IN-2014-JR162107-41675 2014-02-05 2014-02-07  Second Class   
2   25330    IN-2014-CR127307-41929 2014-10-17 2014-10-18   First Class   
3   13524   ES-2014-KM1637548-41667 2014-01-28 2014-01-30   First Class   
4   47221   SG-2014-RH9495111-41948 2014-11-05 2014-11-06      Same Day   

    Customer ID     Customer Name      Segment Postal Code           City  \
0  AB-100151402     Aaron Bergman     Consumer     73120.0  Oklahoma City   
1     JR-162107     Justin Ritter    Corporate         nan     Wollongong   
2     CR-127307      Craig Reiter     Consumer         nan       Brisbane   
3    KM-1637548  Katherine Murray  Home Office         nan         Berlin   
4    RH-9495111       Rick Hansen     Consumer         nan          Dakar   

   ...    Profit Shipping Cost Order Priority Total_Order_Value profit margin  \
0  ...   62.1544         40.77           High           443.960     28.000000   
1  ... -288.7650        923.63       Critical         33384.555     -7.784693   
2  ...  919.9710        915.49         Medium         46576.539     17.776630   
3  ...  -96.5400        910.16         Medium         14462.550     -3.337586   
4  ...  311.5200        903.04       Critical         22663.680     10.996272   

  price_after_discount Shipping_cost_per_unit order_year  order_month  \
0             221.9800              20.385000       2014           11   
1            3338.4555             102.625556       2014            2   
2            4657.6539             101.721111       2014           10   
3            2603.2590             182.032000       2014            1   
4            2832.9600             112.880000       2014           11   

   order_month_name  
0          November  
1          February  
2           October  
3           January  
4          November  

[5 rows x 31 columns]

### Day of the week on which the order was placed

In [141]:
data['order_day']=data['Order Date'].dt.day
data['order_day_name']=data['Order Date'].dt.day_name()
data.head()

Row ID                  Order ID Order Date  Ship Date     Ship Mode  \
0   40098  CA-2014-AB10015140-41954 2014-11-11 2014-11-13   First Class   
1   26341    IN-2014-JR162107-41675 2014-02-05 2014-02-07  Second Class   
2   25330    IN-2014-CR127307-41929 2014-10-17 2014-10-18   First Class   
3   13524   ES-2014-KM1637548-41667 2014-01-28 2014-01-30   First Class   
4   47221   SG-2014-RH9495111-41948 2014-11-05 2014-11-06      Same Day   

    Customer ID     Customer Name      Segment Postal Code           City  \
0  AB-100151402     Aaron Bergman     Consumer     73120.0  Oklahoma City   
1     JR-162107     Justin Ritter    Corporate         nan     Wollongong   
2     CR-127307      Craig Reiter     Consumer         nan       Brisbane   
3    KM-1637548  Katherine Murray  Home Office         nan         Berlin   
4    RH-9495111       Rick Hansen     Consumer         nan          Dakar   

   ... Order Priority Total_Order_Value profit margin price_after_discount  \
0  ...           High           443.960     28.000000             221.9800   
1  ...       Critical         33384.555     -7.784693            3338.4555   
2  ...         Medium         46576.539     17.776630            4657.6539   
3  ...         Medium         14462.550     -3.337586            2603.2590   
4  ...       Critical         22663.680     10.996272            2832.9600   

  Shipping_cost_per_unit order_year order_month order_month_name  order_day  \
0              20.385000       2014          11         November         11   
1             102.625556       2014           2         February          5   
2             101.721111       2014          10          October         17   
3             182.032000       2014           1          January         28   
4             112.880000       2014          11         November          5   

   order_day_name  
0         Tuesday  
1       Wednesday  
2          Friday  
3         Tuesday  
4       Wednesday  

[5 rows x 33 columns]

In [142]:
# Sample data for more feature engineering

In [143]:
data1 = { 'Timestamp': ['2024-07-24 14:45:00', '2024-07-25 16:30:00',
                      '2024-07-26 08:20:00', '2024-07-27 12:15:00'] }
df = pd.DataFrame(data1)
df

Timestamp
0  2024-07-24 14:45:00
1  2024-07-25 16:30:00
2  2024-07-26 08:20:00
3  2024-07-27 12:15:00

In [144]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 1 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   Timestamp  4 non-null      object
dtypes: object(1)
memory usage: 160.0+ bytes


In [145]:
# df['Timestamp']=(df['Timestamp']).astype('datetime64[ns]')
# df.info()

In [146]:
df['Timestamp']=pd.to_datetime(df['Timestamp'])
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 1 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   Timestamp  4 non-null      datetime64[ns]
dtypes: datetime64[ns](1)
memory usage: 160.0 bytes


In [147]:
df

Timestamp
0 2024-07-24 14:45:00
1 2024-07-25 16:30:00
2 2024-07-26 08:20:00
3 2024-07-27 12:15:00

In [148]:
df['date']=df['Timestamp'].dt.date
df['year']=df['Timestamp'].dt.year
df['month']=df['Timestamp'].dt.month
df['hour']=df['Timestamp'].dt.hour
df['minute']=df['Timestamp'].dt.minute
df['Seconds']=df['Timestamp'].dt.second
df

Timestamp        date  year  month  hour  minute  Seconds
0 2024-07-24 14:45:00  2024-07-24  2024      7    14      45        0
1 2024-07-25 16:30:00  2024-07-25  2024      7    16      30        0
2 2024-07-26 08:20:00  2024-07-26  2024      7     8      20        0
3 2024-07-27 12:15:00  2024-07-27  2024      7    12      15        0

# date_range

In [149]:
ds=pd.date_range(start='2023-01-01',end='2023-12-31',freq='D')
ds

DatetimeIndex(['2023-01-01', '2023-01-02', '2023-01-03', '2023-01-04',
               '2023-01-05', '2023-01-06', '2023-01-07', '2023-01-08',
               '2023-01-09', '2023-01-10',
               ...
               '2023-12-22', '2023-12-23', '2023-12-24', '2023-12-25',
               '2023-12-26', '2023-12-27', '2023-12-28', '2023-12-29',
               '2023-12-30', '2023-12-31'],
              dtype='datetime64[ns]', length=365, freq='D')

In [150]:
ds=pd.date_range(start='2023-01-01',end='2023-12-31',freq='M')
ds

DatetimeIndex(['2023-01-31', '2023-02-28', '2023-03-31', '2023-04-30',
               '2023-05-31', '2023-06-30', '2023-07-31', '2023-08-31',
               '2023-09-30', '2023-10-31', '2023-11-30', '2023-12-31'],
              dtype='datetime64[ns]', freq='M')

In [151]:
ds=pd.date_range(start='2023-01-01',end='2023-12-31',periods=10)
ds

DatetimeIndex(['2023-01-01 00:00:00', '2023-02-10 10:40:00',
               '2023-03-22 21:20:00', '2023-05-02 08:00:00',
               '2023-06-11 18:40:00', '2023-07-22 05:20:00',
               '2023-08-31 16:00:00', '2023-10-11 02:40:00',
               '2023-11-20 13:20:00', '2023-12-31 00:00:00'],
              dtype='datetime64[ns]', freq=None)

### Create a new column for the order delivery time (difference between 'Ship Date' and 'Order Date').

In [152]:
data['Delivery_time']=(data['Ship Date']-data['Order Date']).dt.days
data.head()

Row ID                  Order ID Order Date  Ship Date     Ship Mode  \
0   40098  CA-2014-AB10015140-41954 2014-11-11 2014-11-13   First Class   
1   26341    IN-2014-JR162107-41675 2014-02-05 2014-02-07  Second Class   
2   25330    IN-2014-CR127307-41929 2014-10-17 2014-10-18   First Class   
3   13524   ES-2014-KM1637548-41667 2014-01-28 2014-01-30   First Class   
4   47221   SG-2014-RH9495111-41948 2014-11-05 2014-11-06      Same Day   

    Customer ID     Customer Name      Segment Postal Code           City  \
0  AB-100151402     Aaron Bergman     Consumer     73120.0  Oklahoma City   
1     JR-162107     Justin Ritter    Corporate         nan     Wollongong   
2     CR-127307      Craig Reiter     Consumer         nan       Brisbane   
3    KM-1637548  Katherine Murray  Home Office         nan         Berlin   
4    RH-9495111       Rick Hansen     Consumer         nan          Dakar   

   ... Total_Order_Value profit margin price_after_discount  \
0  ...           443.960     28.000000             221.9800   
1  ...         33384.555     -7.784693            3338.4555   
2  ...         46576.539     17.776630            4657.6539   
3  ...         14462.550     -3.337586            2603.2590   
4  ...         22663.680     10.996272            2832.9600   

  Shipping_cost_per_unit order_year order_month order_month_name order_day  \
0              20.385000       2014          11         November        11   
1             102.625556       2014           2         February         5   
2             101.721111       2014          10          October        17   
3             182.032000       2014           1          January        28   
4             112.880000       2014          11         November         5   

   order_day_name  Delivery_time  
0         Tuesday              2  
1       Wednesday              2  
2          Friday              1  
3         Tuesday              2  
4       Wednesday              1  

[5 rows x 34 columns]

### Display current date

In [153]:
import datetime
print(datetime.datetime.now())

2024-08-22 07:04:14.833937


### Create a feature that calculates the age of an order in days from the order date to today’s date.

In [154]:
data['Order age']=(datetime.datetime.now()-data['Order Date']).dt.days
data.head()

Row ID                  Order ID Order Date  Ship Date     Ship Mode  \
0   40098  CA-2014-AB10015140-41954 2014-11-11 2014-11-13   First Class   
1   26341    IN-2014-JR162107-41675 2014-02-05 2014-02-07  Second Class   
2   25330    IN-2014-CR127307-41929 2014-10-17 2014-10-18   First Class   
3   13524   ES-2014-KM1637548-41667 2014-01-28 2014-01-30   First Class   
4   47221   SG-2014-RH9495111-41948 2014-11-05 2014-11-06      Same Day   

    Customer ID     Customer Name      Segment Postal Code           City  \
0  AB-100151402     Aaron Bergman     Consumer     73120.0  Oklahoma City   
1     JR-162107     Justin Ritter    Corporate         nan     Wollongong   
2     CR-127307      Craig Reiter     Consumer         nan       Brisbane   
3    KM-1637548  Katherine Murray  Home Office         nan         Berlin   
4    RH-9495111       Rick Hansen     Consumer         nan          Dakar   

   ... profit margin price_after_discount Shipping_cost_per_unit order_year  \
0  ...     28.000000             221.9800              20.385000       2014   
1  ...     -7.784693            3338.4555             102.625556       2014   
2  ...     17.776630            4657.6539             101.721111       2014   
3  ...     -3.337586            2603.2590             182.032000       2014   
4  ...     10.996272            2832.9600             112.880000       2014   

  order_month order_month_name order_day order_day_name  Delivery_time  \
0          11         November        11        Tuesday              2   
1           2         February         5      Wednesday              2   
2          10          October        17         Friday              1   
3           1          January        28        Tuesday              2   
4          11         November         5      Wednesday              1   

   Order age  
0       3572  
1       3851  
2       3597  
3       3859  
4       3578  

[5 rows x 35 columns]

In [155]:
data.columns

Index(['Row ID', 'Order ID', 'Order Date', 'Ship Date', 'Ship Mode',
       'Customer ID', 'Customer Name', 'Segment', 'Postal Code', 'City',
       'State', 'Country', 'Region', 'Market', 'Product ID', 'Category',
       'Sub-Category', 'Product Name', 'Sales', 'Quantity', 'Discount',
       'Profit', 'Shipping Cost', 'Order Priority', 'Total_Order_Value',
       'profit margin', 'price_after_discount', 'Shipping_cost_per_unit',
       'order_year', 'order_month', 'order_month_name', 'order_day',
       'order_day_name', 'Delivery_time', 'Order age'],
      dtype='object')

### Average Shipping Delay for Orders Placed in December

In [156]:
december_orders = data[data['Order Date'].dt.month == 12]
december_orders.head()

Row ID                  Order ID Order Date  Ship Date       Ship Mode  \
12   45794   SA-2012-MM7260110-41269 2012-12-26 2012-12-28    Second Class   
19   49463   TZ-2015-RH9555129-42343 2015-12-05 2015-12-07    Second Class   
26   30191  IN-2012-PB19210127-41259 2012-12-16 2012-12-19     First Class   
30     220   US-2012-RR1952536-41270 2012-12-27 2012-12-29    Second Class   
32   40977  CA-2013-AH10030140-41635 2013-12-27 2013-12-31  Standard Class   

     Customer ID    Customer Name    Segment Postal Code           City  ...  \
12    MM-7260110  Magdelene Morse   Consumer         nan          Jizan  ...   
19    RH-9555129  Ritsa Hightower   Consumer         nan         Uvinza  ...   
26   PB-19210127   Phillip Breyer  Corporate         nan         Taipei  ...   
30    RR-1952536        Rick Reed  Corporate         nan  Santo Domingo  ...   
32  AH-100301404    Aaron Hawkins  Corporate     94122.0  San Francisco  ...   

   profit margin price_after_discount Shipping_cost_per_unit order_year  \
12     43.997616             2616.960             208.102500       2012   
19     23.998311             3409.740             127.230000       2015   
26     41.999580             1715.160             362.785000       2012   
30     -8.750236             1357.312             140.811200       2012   
32     11.250000              534.528               5.082222       2013   

   order_month order_month_name order_day order_day_name  Delivery_time  \
12          12         December        26      Wednesday              2   
19          12         December         5       Saturday              2   
26          12         December        16         Sunday              3   
30          12         December        27       Thursday              2   
32          12         December        27         Friday              4   

    Order age  
12       4257  
19       3183  
26       4267  
30       4256  
32       3891  

[5 rows x 35 columns]

In [157]:
avg_delay = (december_orders['Ship Date'] - december_orders['Order Date']).mean()
print(avg_delay)

3 days 22:53:29.333546428


# Replace

In [158]:
data['Ship Mode'].unique()

array(['First Class', 'Second Class', 'Same Day', 'Standard Class'],
      dtype=object)

### How can you replace all instances of the word "Standard Class" with "Std" in the 'Ship Mode' column

In [159]:
data['Ship Mode']=data['Ship Mode'].replace('Standard Class','Std')

In [160]:
data['Ship Mode'].unique()

array(['First Class', 'Second Class', 'Same Day', 'Std'], dtype=object)

In [161]:
data['Order Priority'].unique()

array(['High', 'Critical', 'Medium', 'Low'], dtype=object)

### Replace 'Low' value in 'Order Priority' column with 'Standard' .

In [162]:
data['Order Priority']=data['Order Priority'].replace('Low','Standard')
data['Order Priority'].unique()

array(['High', 'Critical', 'Medium', 'Standard'], dtype=object)

# Filter& Sort

In [163]:
data.head()

Row ID                  Order ID Order Date  Ship Date     Ship Mode  \
0   40098  CA-2014-AB10015140-41954 2014-11-11 2014-11-13   First Class   
1   26341    IN-2014-JR162107-41675 2014-02-05 2014-02-07  Second Class   
2   25330    IN-2014-CR127307-41929 2014-10-17 2014-10-18   First Class   
3   13524   ES-2014-KM1637548-41667 2014-01-28 2014-01-30   First Class   
4   47221   SG-2014-RH9495111-41948 2014-11-05 2014-11-06      Same Day   

    Customer ID     Customer Name      Segment Postal Code           City  \
0  AB-100151402     Aaron Bergman     Consumer     73120.0  Oklahoma City   
1     JR-162107     Justin Ritter    Corporate         nan     Wollongong   
2     CR-127307      Craig Reiter     Consumer         nan       Brisbane   
3    KM-1637548  Katherine Murray  Home Office         nan         Berlin   
4    RH-9495111       Rick Hansen     Consumer         nan          Dakar   

   ... profit margin price_after_discount Shipping_cost_per_unit order_year  \
0  ...     28.000000             221.9800              20.385000       2014   
1  ...     -7.784693            3338.4555             102.625556       2014   
2  ...     17.776630            4657.6539             101.721111       2014   
3  ...     -3.337586            2603.2590             182.032000       2014   
4  ...     10.996272            2832.9600             112.880000       2014   

  order_month order_month_name order_day order_day_name  Delivery_time  \
0          11         November        11        Tuesday              2   
1           2         February         5      Wednesday              2   
2          10          October        17         Friday              1   
3           1          January        28        Tuesday              2   
4          11         November         5      Wednesday              1   

   Order age  
0       3572  
1       3851  
2       3597  
3       3859  
4       3578  

[5 rows x 35 columns]

In [164]:
df= data[data['Segment']=='Consumer']
df.head()

Row ID                  Order ID Order Date  Ship Date    Ship Mode  \
0   40098  CA-2014-AB10015140-41954 2014-11-11 2014-11-13  First Class   
2   25330    IN-2014-CR127307-41929 2014-10-17 2014-10-18  First Class   
4   47221   SG-2014-RH9495111-41948 2014-11-05 2014-11-06     Same Day   
6   30570   IN-2012-TS2134092-41219 2012-11-06 2012-11-08  First Class   
7   31192   IN-2013-MB1808592-41378 2013-04-14 2013-04-18          Std   

    Customer ID  Customer Name   Segment Postal Code           City  ...  \
0  AB-100151402  Aaron Bergman  Consumer     73120.0  Oklahoma City  ...   
2     CR-127307   Craig Reiter  Consumer         nan       Brisbane  ...   
4    RH-9495111    Rick Hansen  Consumer         nan          Dakar  ...   
6    TS-2134092  Toby Swindell  Consumer         nan        Porirua  ...   
7    MB-1808592     Mick Brown  Consumer         nan       Hamilton  ...   

  profit margin price_after_discount Shipping_cost_per_unit order_year  \
0     28.000000             221.9800              20.385000       2014   
2     17.776630            4657.6539             101.721111       2014   
4     10.996272            2832.9600             112.880000       2014   
6     30.999737            1822.0800             223.692500       2012   
7     18.999245            5244.8400             146.396667       2013   

  order_month order_month_name order_day order_day_name  Delivery_time  \
0          11         November        11        Tuesday              2   
2          10          October        17         Friday              1   
4          11         November         5      Wednesday              1   
6          11         November         6        Tuesday              2   
7           4            April        14         Sunday              4   

   Order age  
0       3572  
2       3597  
4       3578  
6       4307  
7       4148  

[5 rows x 35 columns]

In [165]:
data['Profit'].unique()

array([  62.1544, -288.765 ,  919.971 , ...,   22.0472,  -49.572 ,
         -6.456 ])

## Filter the rows where profit less than 0

In [166]:
pf_low=data[data['Profit']<0]
pf_low.head()

Row ID                  Order ID Order Date  Ship Date     Ship Mode  \
1    26341    IN-2014-JR162107-41675 2014-02-05 2014-02-07  Second Class   
3    13524   ES-2014-KM1637548-41667 2014-01-28 2014-01-30   First Class   
16   39519  CA-2012-AB10015140-40958 2012-02-19 2012-02-25           Std   
30     220   US-2012-RR1952536-41270 2012-12-27 2012-12-29  Second Class   
40   21316   ID-2014-JB1600059-41873 2014-08-22 2014-08-26           Std   

     Customer ID     Customer Name      Segment Postal Code           City  \
1      JR-162107     Justin Ritter    Corporate         nan     Wollongong   
3     KM-1637548  Katherine Murray  Home Office         nan         Berlin   
16  AB-100151402     Aaron Bergman     Consumer     76017.0      Arlington   
30    RR-1952536         Rick Reed    Corporate         nan  Santo Domingo   
40    JB-1600059         Joy Bell-     Consumer         nan        Mataram   

    ... profit margin price_after_discount Shipping_cost_per_unit order_year  \
1   ...     -7.784693          3338.455500             102.625556       2014   
3   ...     -3.337586          2603.259000             182.032000       2014   
16  ...    -20.000000            10.099200               0.985000       2012   
30  ...     -8.750236          1357.312000             140.811200       2012   
40  ...     -2.412169          2656.494846             110.145000       2014   

   order_month order_month_name order_day order_day_name  Delivery_time  \
1            2         February         5      Wednesday              2   
3            1          January        28        Tuesday              2   
16           2         February        19         Sunday              6   
30          12         December        27       Thursday              2   
40           8           August        22         Friday              4   

    Order age  
1        3851  
3        3859  
16       4568  
30       4256  
40       3653  

[5 rows x 35 columns]

## Filter rows where the 'Profit' is negative and the 'Quantity' is greater than 10?

In [167]:
filter_rows=data[(data['Profit']<0) & (data['Quantity']>10)]
filter_rows.head()

Row ID                  Order ID Order Date  Ship Date     Ship Mode  \
109   11743   IT-2014-VM2168591-41711 2014-03-13 2014-03-13      Same Day   
289   43395  TU-2013-SF10200134-41417 2013-05-23 2013-05-26   First Class   
298   16456  IT-2013-JD16060120-41515 2013-08-29 2013-09-01   First Class   
362   29513   ID-2014-SN2056097-41896 2014-09-14 2014-09-14      Same Day   
368   29552   IN-2015-SV2078559-42110 2015-04-16 2015-04-18  Second Class   

     Customer ID     Customer Name      Segment Postal Code            City  \
109   VM-2168591   Valerie Mitchum  Home Office         nan  Bergen op Zoom   
289  SF-10200134      Sarah Foster     Consumer         nan           Izmir   
298  JD-16060120      Julia Dunbar     Consumer         nan       Santander   
362   SN-2056097      Skye Norling  Home Office         nan          Lahore   
368   SV-2078559  Stewart Visinsky     Consumer         nan           Medan   

     ... profit margin price_after_discount Shipping_cost_per_unit order_year  \
109  ...    -86.008600          1285.432500              47.353636       2014   
289  ...   -142.537313           373.699200              25.980714       2013   
298  ...    -10.021090          1253.369700              32.632727       2013   
362  ...    -86.004621          1749.577500              30.281818       2014   
368  ...     -7.231295          1496.559471              30.193636       2015   

    order_month order_month_name order_day order_day_name  Delivery_time  \
109           3            March        13       Thursday              0   
289           5              May        23       Thursday              3   
298           8           August        29       Thursday              3   
362           9        September        14         Sunday              0   
368           4            April        16       Thursday              2   

     Order age  
109       3815  
289       4109  
298       4011  
362       3630  
368       3416  

[5 rows x 35 columns]

## Calculate the average sales of these filtered rows

In [168]:
filter_rows['Sales'].mean()

650.2714193333333

## Modify the 'Order Priority' column to label all orders with a 'Profit' below zero as "Urgent"?

In [174]:
data.loc[data['Profit']<0, 'Order Priority']='Urgent'


In [176]:
data['Order Priority'].unique()

array(['High', 'Urgent', 'Medium', 'Critical', 'Standard'], dtype=object)

In [179]:
data.head()

Row ID                  Order ID Order Date  Ship Date     Ship Mode  \
0   40098  CA-2014-AB10015140-41954 2014-11-11 2014-11-13   First Class   
1   26341    IN-2014-JR162107-41675 2014-02-05 2014-02-07  Second Class   
2   25330    IN-2014-CR127307-41929 2014-10-17 2014-10-18   First Class   
3   13524   ES-2014-KM1637548-41667 2014-01-28 2014-01-30   First Class   
4   47221   SG-2014-RH9495111-41948 2014-11-05 2014-11-06      Same Day   

    Customer ID     Customer Name      Segment Postal Code           City  \
0  AB-100151402     Aaron Bergman     Consumer     73120.0  Oklahoma City   
1     JR-162107     Justin Ritter    Corporate         nan     Wollongong   
2     CR-127307      Craig Reiter     Consumer         nan       Brisbane   
3    KM-1637548  Katherine Murray  Home Office         nan         Berlin   
4    RH-9495111       Rick Hansen     Consumer         nan          Dakar   

   ... profit margin price_after_discount Shipping_cost_per_unit order_year  \
0  ...     28.000000             221.9800              20.385000       2014   
1  ...     -7.784693            3338.4555             102.625556       2014   
2  ...     17.776630            4657.6539             101.721111       2014   
3  ...     -3.337586            2603.2590             182.032000       2014   
4  ...     10.996272            2832.9600             112.880000       2014   

  order_month order_month_name order_day order_day_name  Delivery_time  \
0          11         November        11        Tuesday              2   
1           2         February         5      Wednesday              2   
2          10          October        17         Friday              1   
3           1          January        28        Tuesday              2   
4          11         November         5      Wednesday              1   

   Order age  
0       3572  
1       3851  
2       3597  
3       3859  
4       3578  

[5 rows x 35 columns]

In [180]:
data['Order Priority']

0            High
1          Urgent
2          Medium
3          Urgent
4        Critical
           ...   
51285      Medium
51286        High
51287        High
51288      Medium
51289        High
Name: Order Priority, Length: 51290, dtype: object

# How  many rows have 'Urgent'as the Order Priority

In [190]:
len(data[data['Order Priority']=='Urgent'])

12544

In [187]:
data[data['Order Priority']=='Urgent'].shape[0]

(12544, 35)

## Filter data where category is 'Furniture' and profit is greater than 1000

In [83]:
data['Category'].unique()

array(['Technology', 'Furniture', 'Office Supplies'], dtype=object)

In [181]:
fur=data[(data['Category']=='Furniture')& (data['Profit']>1000)]
fur.head()

Row ID                  Order ID Order Date  Ship Date     Ship Mode  \
42   25795   IN-2015-VG2180558-42273 2015-09-26 2015-09-28  Second Class   
66   16653  IT-2015-DK13090120-42345 2015-12-07 2015-12-08   First Class   
70   21191   IN-2014-AS1022527-41797 2014-06-07 2014-06-09   First Class   
72   49085    CG-2014-BW106533-41727 2014-03-29 2014-03-31  Second Class   
88   27962   IN-2014-BK1126011-41940 2014-10-28 2014-10-31   First Class   

    Customer ID      Customer Name    Segment Postal Code                City  \
42   VG-2180558        Vivek Grady  Corporate         nan  Thiruvananthapuram   
66  DK-13090120          Dave Kipp   Consumer         nan             Seville   
70   AS-1022527  Alan Schoenberger  Corporate         nan               Jinan   
72    BW-106533      Barry Weirich   Consumer         nan             Kananga   
88   BK-1126011     Berenike Kampe   Consumer         nan               Tongi   

    ... profit margin price_after_discount Shipping_cost_per_unit order_year  \
42  ...     36.998555              5667.87              50.642308       2015   
66  ...     47.994790              2188.05             118.782000       2015   
70  ...     31.994142              3298.26              84.194286       2014   
72  ...     40.000000              3808.80              73.516250       2014   
88  ...     31.993627              3728.43              61.478889       2014   

   order_month order_month_name order_day order_day_name  Delivery_time  \
42           9        September        26       Saturday              2   
66          12         December         7         Monday              1   
70           6             June         7       Saturday              2   
72           3            March        29       Saturday              2   
88          10          October        28        Tuesday              3   

    Order age  
42       3253  
66       3181  
70       3729  
72       3799  
88       3586  

[5 rows x 35 columns]

## Filter rows where product name contain word 'phone'




In [184]:
ph=data[data['Product Name'].str.contains('Phone')]
ph.head()

Row ID                 Order ID Order Date  Ship Date     Ship Mode  \
2    25330   IN-2014-CR127307-41929 2014-10-17 2014-10-18   First Class   
3    13524  ES-2014-KM1637548-41667 2014-01-28 2014-01-30   First Class   
5    22732   IN-2014-JM156557-41818 2014-06-28 2014-07-01  Second Class   
12   45794  SA-2012-MM7260110-41269 2012-12-26 2012-12-28  Second Class   
24    1570  US-2015-NP1832582-42216 2015-07-31 2015-08-01   First Class   

   Customer ID     Customer Name      Segment Postal Code      City  ...  \
2    CR-127307      Craig Reiter     Consumer         nan  Brisbane  ...   
3   KM-1637548  Katherine Murray  Home Office         nan    Berlin  ...   
5    JM-156557       Jim Mitchum    Corporate         nan    Sydney  ...   
12  MM-7260110   Magdelene Morse     Consumer         nan     Jizan  ...   
24  NP-1832582      Naresj Patel     Consumer         nan    Juárez  ...   

   profit margin price_after_discount Shipping_cost_per_unit order_year  \
2      17.776630            4657.6539             101.721111       2014   
3      -3.337586            2603.2590             182.032000       2014   
5      26.662999            2576.4075             179.470000       2014   
12     43.997616            2616.9600             208.102500       2012   
24     25.995425            1713.8400             182.242000       2015   

   order_month order_month_name order_day order_day_name  Delivery_time  \
2           10          October        17         Friday              1   
3            1          January        28        Tuesday              2   
5            6             June        28       Saturday              3   
12          12         December        26      Wednesday              2   
24           7             July        31         Friday              1   

    Order age  
2        3597  
3        3859  
5        3708  
12       4257  
24       3310  

[5 rows x 35 columns]

In [86]:
data['Product Name'].unique()

array(['Samsung Convoy 3', 'Novimex Executive Leather Armchair, Black',
       'Nokia Smart Phone, with Caller ID', ...,
       'Belkin 8 Outlet SurgeMaster II Gold Surge Protector with Phone Protection',
       'Hewlett-Packard Deskjet D4360 Printer', 'Cisco SPA 501G IP Phone'],
      dtype=object)

In [87]:
data['Product Name'].value_counts()

Product Name
Staples                                                 227
Cardinal Index Tab, Clear                                92
Eldon File Cart, Single Width                            90
Rogers File Cart, Single Width                           84
Ibico Index Tab, Clear                                   83
                                                       ... 
Lexmark X 9575 Professional All-in-One Color Printer      1
NETGEAR RangeMax WNR1000 Wireless Router                  1
4009 Highlighters                                         1
Xerox 1938                                                1
Cisco SPA 501G IP Phone                                   1
Name: count, Length: 3788, dtype: int64

## Filter the dataset to include only rows where the 'Order Date' is between '2013-01-01' and '2014-12-31'?

In [207]:
data.describe(include='datetime')

Order Date                      Ship Date
count                          51290                          51290
mean   2014-05-11 21:26:49.155780608  2014-05-15 20:42:42.745174528
min              2012-01-01 00:00:00            2012-01-03 00:00:00
25%              2013-06-19 00:00:00            2013-06-23 00:00:00
50%              2014-07-08 00:00:00            2014-07-12 00:00:00
75%              2015-05-22 00:00:00            2015-05-26 00:00:00
max              2015-12-31 00:00:00            2016-01-07 00:00:00

# Sorting

In [193]:
data.columns

Index(['Row ID', 'Order ID', 'Order Date', 'Ship Date', 'Ship Mode',
       'Customer ID', 'Customer Name', 'Segment', 'Postal Code', 'City',
       'State', 'Country', 'Region', 'Market', 'Product ID', 'Category',
       'Sub-Category', 'Product Name', 'Sales', 'Quantity', 'Discount',
       'Profit', 'Shipping Cost', 'Order Priority', 'Total_Order_Value',
       'profit margin', 'price_after_discount', 'Shipping_cost_per_unit',
       'order_year', 'order_month', 'order_month_name', 'order_day',
       'order_day_name', 'Delivery_time', 'Order age'],
      dtype='object')

In [194]:
data['Sales'].sort_values(ascending=False)

47114    22638.480
49003    17499.950
44248    13999.960
49940    11199.968
25238    10499.970
           ...    
24977        0.876
26588        0.852
33493        0.836
45391        0.556
51260        0.444
Name: Sales, Length: 51290, dtype: float64

In [202]:
data[['Product Name','Sales']].sort_values(by='Sales',ascending=False)

Product Name      Sales
47114  Cisco TelePresence System EX90 Videoconferenci...  22638.480
49003              Canon imageCLASS 2200 Advanced Copier  17499.950
44248              Canon imageCLASS 2200 Advanced Copier  13999.960
49940              Canon imageCLASS 2200 Advanced Copier  11199.968
25238              Canon imageCLASS 2200 Advanced Copier  10499.970
...                                                  ...        ...
24977                                  Acco 3-Hole Punch      0.876
26588                      Avery Round Ring Poly Binders      0.852
33493  Avery Durable Slant Ring Binders With Label Ho...      0.836
45391           Acco Suede Grain Vinyl Round Ring Binder      0.556
51260  Hoover Replacement Belt for Commercial Guardsm...      0.444

[51290 rows x 2 columns]

## Sort the dataset first by 'Category' in ascending order and then by 'Profit' in descending order?

In [206]:
data[['Category','Profit']].sort_values(by=['Category','Profit'],ascending=[True,False])

Category     Profit
334     Furniture  2316.5100
42      Furniture  2097.0300
192     Furniture  2071.4400
118     Furniture  1868.1300
940     Furniture  1622.8800
...           ...        ...
362    Technology -3009.4350
47527  Technology -3399.9800
23110  Technology -3839.9904
6591   Technology -4088.3760
11755  Technology -6599.9780

[51290 rows x 2 columns]

## How can you sort the dataset by 'Quantity' in descending order and then select the top 10 records?

In [209]:
data.sort_values(by='Quantity',ascending=False).head(10)

Row ID                  Order ID Order Date  Ship Date     Ship Mode  \
7971    43213    CG-2015-JR570033-42272 2015-09-25 2015-09-30           Std   
17405   19420  IT-2015-MG18205139-42242 2015-08-26 2015-08-29  Second Class   
40522   34244  CA-2015-MP18175140-42353 2015-12-15 2015-12-19           Std   
15388   26869    IN-2012-MW182357-41052 2012-05-23 2012-05-27           Std   
33134   43138    NI-2014-PR888095-41982 2014-12-09 2014-12-11   First Class   
13398   32726  US-2013-DK12835140-41578 2013-10-31 2013-11-04  Second Class   
17675   47093     AU-2015-DA34508-42355 2015-12-17 2015-12-22           Std   
10107   19306  ES-2013-SC20260139-41625 2013-12-17 2013-12-21  Second Class   
2680    49660  TU-2013-TA11385134-41326 2013-02-21 2013-02-27           Std   
31397   49851   PL-2015-BD1725103-42217 2015-08-01 2015-08-08           Std   

        Customer ID     Customer Name      Segment Postal Code           City  \
7971      JR-570033    Jocasta Rupert     Consumer         nan       Kinshasa   
17405   MG-18205139   Mitch Gastineau    Corporate         nan  Wolverhampton   
40522  MP-181751404    Mike Pelletier  Home Office     92024.0      San Diego   
15388     MW-182357  Mitch Willingham    Corporate         nan      Mount Isa   
33134     PR-888095      Patrick Ryan     Consumer         nan        Calabar   
13398  DK-128351404   Damala Kotsonis    Corporate     93905.0        Salinas   
17675      DA-34508     Dianna Arnett  Home Office         nan           Linz   
10107   SC-20260139       Scott Cohen    Corporate         nan         Bolton   
2680    TA-11385134      Tom Ashbrook  Home Office         nan        Silivri   
31397    BD-1725103  Bruce Degenhardt     Consumer         nan           Lodz   

       ... profit margin price_after_discount Shipping_cost_per_unit  \
7971   ...     34.989083             769.4400               3.040714   
17405  ...     -4.339347             180.8982               1.125714   
40522  ...     35.000000              51.9680               0.380000   
15388  ...     33.201058             171.4608               1.358571   
33134  ...    -90.775681               6.0102               0.301429   
13398  ...     46.000000             427.4200               3.018571   
17675  ...      8.984831             719.8800               1.100714   
10107  ...     45.000000             403.2000               2.317857   
2680   ...   -125.004251             395.2032               7.707857   
31397  ...     40.449438              74.7600               0.350000   

      order_year order_month order_month_name order_day order_day_name  \
7971        2015           9        September        25         Friday   
17405       2015           8           August        26      Wednesday   
40522       2015          12         December        15        Tuesday   
15388       2012           5              May        23      Wednesday   
33134       2014          12         December         9        Tuesday   
13398       2013          10          October        31       Thursday   
17675       2015          12         December        17       Thursday   
10107       2013          12         December        17        Tuesday   
2680        2013           2         February        21       Thursday   
31397       2015           8           August         1       Saturday   

       Delivery_time  Order age  
7971               5       3254  
17405              3       3284  
40522              4       3173  
15388              4       4474  
33134              2       3544  
13398              4       3948  
17675              5       3171  
10107              4       3901  
2680               6       4200  
31397              7       3309  

[10 rows x 35 columns]

## Sort the dataset by 'Order Date' in ascending order?

In [210]:
data.sort_values(by='Order Date',ascending=True)

Row ID                  Order ID Order Date  Ship Date     Ship Mode  \
22950   22253    IN-2012-JH159857-40909 2012-01-01 2012-01-08           Std   
43610   22254    IN-2012-JH159857-40909 2012-01-01 2012-01-08           Std   
25078   48883     HU-2012-AT73557-40909 2012-01-01 2012-01-05  Second Class   
31870   22255    IN-2012-JH159857-40909 2012-01-01 2012-01-08           Std   
31583   11731  IT-2012-EM14140124-40909 2012-01-01 2012-01-05  Second Class   
...       ...                       ...        ...        ...           ...   
17379   10625   ES-2015-ES1402048-42369 2015-12-31 2016-01-04           Std   
42870   32203  CA-2015-PO18865140-42369 2015-12-31 2016-01-04           Std   
39873    5029   MX-2015-SC2026082-42369 2015-12-31 2016-01-02  Second Class   
7908    28570  ID-2015-MY18295102-42369 2015-12-31 2016-01-04           Std   
39602    7200   MX-2015-RB1946593-42369 2015-12-31 2015-12-31      Same Day   

        Customer ID      Customer Name      Segment Postal Code  \
22950     JH-159857        Joseph Holt     Consumer         nan   
43610     JH-159857        Joseph Holt     Consumer         nan   
25078      AT-73557      Annie Thurman     Consumer         nan   
31870     JH-159857        Joseph Holt     Consumer         nan   
31583   EM-14140124       Eugene Moren  Home Office         nan   
...             ...                ...          ...         ...   
17379    ES-1402048        Erica Smith     Consumer         nan   
42870  PO-188651406  Patrick O'Donnell     Consumer     10009.0   
39873    SC-2026082        Scott Cohen    Corporate         nan   
7908    MY-18295102    Muhammed Yedwab    Corporate         nan   
39602    RB-1946593       Rick Bensley  Home Office         nan   

                City  ... profit margin price_after_discount  \
22950    Wagga Wagga  ...     29.938687             108.3294   
43610    Wagga Wagga  ...     27.772347              49.7178   
25078       Budapest  ...     44.827586              66.1200   
31870    Wagga Wagga  ...     33.227765             102.3030   
31583      Stockholm  ...    -58.074223              22.4325   
...              ...  ...           ...                  ...   
17379      Bielefeld  ...      4.968944             260.8200   
42870  New York City  ...      3.750000             258.5088   
39873        Durango  ...     17.789757              22.2600   
7908          Manila  ...     -9.236554             444.9939   
39602        Managua  ...      8.000000              15.0000   

      Shipping_cost_per_unit order_year order_month order_month_name  \
22950               3.240000       2012           1          January   
43610               0.900000       2012           1          January   
25078               2.042500       2012           1          January   
31870               0.940000       2012           1          January   
31583               1.606667       2012           1          January   
...                      ...        ...         ...              ...   
17379               2.631667       2015          12         December   
42870               9.567500       2015          12         December   
39873               0.758000       2015          12         December   
7908               10.730000       2015          12         December   
39602               1.169000       2015          12         December   

      order_day order_day_name  Delivery_time  Order age  
22950         1         Sunday              7       4617  
43610         1         Sunday              7       4617  
25078         1         Sunday              4       4617  
31870         1         Sunday              7       4617  
31583         1         Sunday              4       4617  
...         ...            ...            ...        ...  
17379        31       Thursday              4       3157  
42870        31       Thursday              4       3157  
39873        31       Thursday              2       3157  
7908         31       Thursda